core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java

```diff
@@ -134,36 +137,28 @@ public void runPhase(final ModuleFlowProcessingPhaseTemplate template, final Mes
       // TODO MULE-11167 Policies should be non blocking
       if (!enableSourcePolicies) {
         just(templateEvent)
-            .doOnNext(request -> {
-              fireNotification(messageProcessContext.getMessageSource(), request,
-                               messageProcessContext.getFlowConstruct(),
-                               MESSAGE_RECEIVED);
-            })
-            .then(request -> from(template.routeEventAsync(request)))
-            .doOnSuccess(getSuccessConsumer(messageSource, templateEvent, messageProcessContext, phaseResultNotifier, template,
-                                            terminateConsumer))
-            .doOnError(MessagingException.class, me -> {
-              if (me.getCause() instanceof SourceErrorException
-                  && sourceResponseErrorTypeMatcher.match(((SourceErrorException) me.getCause()).getErrorType())) {
-                try {
-                  handleSourceError(exceptionHandler, errorConsumer, (SourceErrorException) me.getCause());
-                  onTerminate(terminateConsumer, left(me.getEvent()));
-                } catch (Exception e) {
-                  onTerminate(terminateConsumer, right(e));
-                }
+            .doOnNext(request -> fireNotification(messageProcessContext.getMessageSource(), request,
+                                                  messageProcessContext.getFlowConstruct(),
+                                                  MESSAGE_RECEIVED))
+            .then(request -> from(template.routeEventAsync(request)).switchIfEmpty(fromCallable(() -> emptyEvent(templateEvent))))
+            .then(onSuccess(messageSource, templateEvent, messageProcessContext, phaseResultNotifier, template,
+                            terminateConsumer))
+            .onErrorResume(SourceErrorException.class,
+                           see -> onSourceException(exceptionHandler, errorHandler, terminateConsumer, see))
+            .onErrorResume(MessagingException.class, me -> {
+              if (me.getCause() instanceof SourceErrorException) {
```

Comment:
Doesn't our error hook map thrown exceptions into MessagingExceptions if they aren't? In that case onErrorResume shouldn't be able to handle other things (except they are put in the stream with error(e) instead of throw e). @dfeist ?